GooglePlayInstant/LoadingScreen/LoadingScreen.cs

```diff
@@ -0,0 +1,121 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections;
+using UnityEngine;
+using UnityEngine.Networking;
+using UnityEngine.SceneManagement;
+using UnityEngine.UI;
+
+namespace GooglePlayInstant.LoadingScreen
+{
+    /// <summary>
+    /// Downloads the AssetBundle available at AssetBundleUrl and updates LoadingBar with its progress
+    /// </summary>
+    public class LoadingScreen : MonoBehaviour
+    {
+        [Tooltip("The url used to fetch the AssetBundle on Start")]
+        public string AssetBundleUrl;
+
+        [Tooltip("The LoadingBar used to indicated download and install progress")]
+        public LoadingBar LoadingBar;
+
+        public RawImage Background;
+
+        private const int MaxAttemptCount = 3;
+        private AssetBundle _bundle;
+        private int _assetBundleRetrievalAttemptCount;
+
+        private IEnumerator Start()
+        {
+            while (_bundle == null && _assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                yield return GetAssetBundle(AssetBundleUrl);
+            }
+
+            if (_bundle == null)
+            {
+                // TODO: Develop UI for when AssetBundle download fails, e.g. a user prompt with a retry button.
+                Debug.LogErrorFormat("Failed to download AssetBundle after {0} attempts.", MaxAttemptCount);
+                yield break;
+            }
+
+            var sceneLoadOperation = SceneManager.LoadSceneAsync(_bundle.GetAllScenePaths()[0]);
+            yield return LoadingBar.FillUntilDone(sceneLoadOperation, LoadingBar.AssetBundleDownloadToInstallRatio, 1f);
+        }
+
+        private IEnumerator GetAssetBundle(string assetBundleUrl)
+        {
+            UnityWebRequest webRequest;
+            var downloadOperation = StartAssetBundleDownload(assetBundleUrl, out webRequest);
+
+            yield return LoadingBar.FillUntilDone(downloadOperation,
+                0f, LoadingBar.AssetBundleDownloadToInstallRatio);
+
+            if (IsFailedRequest(webRequest))
+            {
+                yield return HandleRequestFailed(assetBundleUrl, webRequest);
+            }
+            else
+            {
+                _bundle = DownloadHandlerAssetBundle.GetContent(webRequest);
+            }
+        }
+
+        private IEnumerator HandleRequestFailed(string assetBundleUrl, UnityWebRequest webRequest)
+        {
+            if (_assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                _assetBundleRetrievalAttemptCount++;
+                Debug.LogFormat("Attempt #{0} at downloading AssetBundle...", _assetBundleRetrievalAttemptCount);
+                yield return new WaitForSeconds(2);
+
+                //TODO: revisit this methodology of setting the loading bar
+                LoadingBar.SetProgress(0f);
```

Comment:
I thought we decided that progress should never go backwards?

Reply:
I had planned to update that when we added the retry button, but for now I'll make it resume from it's last maximum fill.